opal/mca/btl/uct/btl_uct_rdma.c

```diff
@@ -0,0 +1,188 @@
+                      uint64_t remote_address, mca_btl_base_registration_handle_t *local_handle,
+                      mca_btl_base_registration_handle_t *remote_handle, size_t size, int flags,
+                      int order, mca_btl_base_rdma_completion_fn_t cbfunc, void *cbcontext, void *cbdata)
+{
+    mca_btl_uct_module_t *uct_module = (mca_btl_uct_module_t *) btl;
+    mca_btl_uct_device_context_t *context = mca_btl_uct_module_get_context (uct_module);
+    mca_btl_uct_uct_completion_t *comp = NULL;
+    ucs_status_t ucs_status;
+    uct_rkey_bundle_t rkey;
+    uct_ep_h ep_handle;
+    int rc;
+
+    if (size >  uct_module->super.btl_put_local_registration_threshold && cbfunc) {
+        comp = mca_btl_uct_uct_completion_alloc (btl, endpoint, local_address, local_handle, cbfunc, cbcontext, cbdata);
+        if (OPAL_UNLIKELY(NULL == comp)) {
+            return OPAL_ERR_OUT_OF_RESOURCE;
+        }
+    }
+
+    rc = mca_btl_uct_get_rkey (uct_module, context, endpoint, remote_handle, &rkey, &ep_handle);
+    if (OPAL_UNLIKELY(OPAL_SUCCESS != rc)) {
+        mca_btl_uct_uct_completion_release (comp);
+        return rc;
+    }
+
+    mca_btl_uct_context_lock (context);
+
+    if (size > uct_module->super.btl_put_local_registration_threshold) {
+        uct_iov_t iov = {.buffer = local_address, .length = size, .stride = 0, .count = 1,
+                         .memh = MCA_BTL_UCT_REG_REMOTE_TO_LOCAL(local_handle)->uct_memh};
+
+        ucs_status = uct_ep_put_zcopy (ep_handle, &iov, 1, remote_address, rkey.rkey, &comp->uct_comp);
+    } else {
+        ucs_status = uct_ep_put_short (ep_handle, local_address, size, remote_address, rkey.rkey);
+    }
+
+    /* go ahead and progress the worker while we have the lock */
+    (void) uct_worker_progress (context->uct_worker);
+
+    mca_btl_uct_context_unlock (context);
+
+    if (UCS_OK == ucs_status && cbfunc) {
+        /* if UCS_OK is returned the callback will never fire so we have to make the callback
+         * ourselves. this callback is possibly being made before the data is visible to the
+         * remote process. */
+        cbfunc (btl, endpoint, local_address, local_handle, cbcontext, cbdata, OPAL_SUCCESS);
+        free (comp);
+    } else if (UCS_INPROGRESS == ucs_status) {
+        ucs_status = UCS_OK;
+    }
+
+    uct_rkey_release (&rkey);
+
+    return OPAL_LIKELY(UCS_OK == ucs_status) ? OPAL_SUCCESS : OPAL_ERR_RESOURCE_BUSY;
+}
+
+int mca_btl_uct_flush (mca_btl_base_module_t *btl, mca_btl_base_endpoint_t *endpoint)
+{
+    mca_btl_uct_module_t *uct_module = (mca_btl_uct_module_t *) btl;
+    ucs_status_t ucs_status;
+
+    /* flush all device contexts */
+    for (int i = 0 ; i < uct_module->uct_worker_count ; ++i) {
+        mca_btl_uct_device_context_t *context = uct_module->contexts + i;
+
+        if (NULL == context->uct_worker) {
+            continue;
+        }
+
+        mca_btl_uct_context_lock (context);
```

Comment:
@hjelmn I think you'll get the better performance if you will use trylock and will allow multiple threads to independently call progress on multiple workers.
One thread s unlikely to fully utilize the NIC.

Reply:
I meant that now it looks like one thread will be held by another one holding the lock of the context 0 for example.
If use trylock - you can skip context 0 and go to context 1

Reply:
If I’m not mistaken, one concerning thing is that if number of contexts is less than number of threads, this flush in one thread can block other threads from posting new requests while waiting for the current work to finish.

Reply:
The problem is the requirements of MPI_Flush. All operations targeting a remote process (or all remote processes) need to be flushed regardless of which thread started it so we can't skip a device context. This is something that the endpoints proposal would be able to help with. In theory flushing everything will have an impact but without many apps it is hard to tell what the real impact is. I can modify the RMA-MT benchmarks to include this scenario and see how we might be able to work around this limitation without endpoints. It might require tweaking the BTL interface (which is fine).

Reply:
But you can still let multiple threads to do the progress
You can skip context, but do multiple iterations on top of `for(i=0; i < num_contexts;i++)` loop to make sure all is done.
I'd guess this would have better perf as multiple threads will progress.

Reply:
Good point. I will make that change.